api/controllers/clients.js

```diff
@@ -0,0 +1,224 @@
+/*
+    TODO: Update account bio / display name
+*/
+
+const { db } = require('./postgresql-utils');
+const { assertBodyKey, assertHeaderKey } = require('./controller-utils');
+
+const hashing = require('../encryption/hashing');
+const a_Encryption = require('../encryption/asymmetric');
+const s_Encryption = require('../encryption/symmetric');
+
+const bcrypt = require('bcrypt');
+const { v1 } = require('uuid');
+
+const TABLE = 'clients';
+const EMAIL = 'email';
+const DISPLAY_NAME = 'display_name';
+const PASSWORD = 'password';
+const BIO = 'bio';
+const API_KEY = 'api_key';
+const PUBLIC_KEY = 'public_key';
+const PRIVATE_KEY = 'private_key';
+const CLIENT_ID = 'client_id';
+
+const NEW_PASSWORD = 'new_password';
+const NEW_EMAIL = 'new_email';
+
+const SAFETRACE_CLIENT_ID = 1;
+
+const validEmailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+
+async function getClientByEmail (email) {
+    const result = await db.query(`SELECT * FROM ${TABLE} WHERE ${EMAIL} = '${email}';`);
+    return result.rows.length > 0 ? result.rows[0] : null;
+}
+
+function checkForEmailAndPassword (request, action) {
+    let email = assertHeaderKey (request, EMAIL, action);
+    let password = assertHeaderKey (request, PASSWORD, action);    
+    return { email, password };
+}
+
+async function assertClientAuth (request, action) {
+    let { email, password } = checkForEmailAndPassword (request, action);
+    function throwAuthError () {
+        throw new Error(`${action} :: Authorization Failed [Invalid Email Or Password]!`);
+    }
+    let client = await getClientByEmail(email);
+    if (!client) 
+        throwAuthError ();
+    if (!(await bcrypt.compare(password, client[PASSWORD])))
```

Comment:
so does the backend see a plaintext copy of the password? Or are we comparing a hash of the password here? Just a small question

Reply:
the plain text password is sent via the request, but in the request header
its then compared to the hashed password in the DB

Reply:
this client sign up and account stuff would be handled on a static page on our website in the future, so it won't be API endpoints

Reply:
yeah but i would argue only the hashed password should ever touch the backend